README.md
# GnuPG puppet module

[![Build Status](https://travis-ci.com/UGent-DICT/golja-gnupg.svg?branch=master)](https://travis-ci.com/UGent-DICT/golja-gnupg)

#### Table of Contents

1. [Overview](##overview)
2. [Installation](##Installation)
3. [Usage - Configuration options and additional functionality](@#usage)
4. [Reference - An under-the-hood peek at what the module is doing and how](##reference)
5. [Limitations - OS compatibility, etc.](##limitations)
6. [Development - Guide for contributing to the module](##development)
7. [License](##license)

## Overview

Install GnuPG on Ubuntu/Debian/RedHat/CentOS/Amazon AMI and manage users public keys.

Tested with Tavis CI

NOTE: For puppet 2.7.x supported module please use version 0.X.X

[![Build Status](https://travis-ci.org/n1tr0g/golja-gnupg.png)](https://travis-ci.org/n1tr0g/golja-gnupg) [![Puppet Forge](http://img.shields.io/puppetforge/v/golja/gnupg.svg)](https://forge.puppetlabs.com/golja/gnupg)

## Installation

     $ puppet module install golja/gnupg

## Usage

#### Install GnuPG package

    include '::gnupg'

#### Add public key 20BC0A86 from PGP server from hkp://pgp.mit.edu/ to user root

```puppet
gnupg_key { 'hkp_server_20BC0A86':
  ensure     => present,
  key_id     => '20BC0A86',
  user       => 'root',
  key_server => 'hkp://pgp.mit.edu/',
  key_type   => public,
}
```

#### Add public key D50582E6 from standard http URI to user foo

```puppet
gnupg_key { 'jenkins_foo_key':
  ensure     => present,
  key_id     => 'D50582E6',
  user       => 'foo',
  key_source => 'http://pkg.jenkins-ci.org/debian/jenkins-ci.org.key',
  key_type   => public,
}
```

#### Add public key D50582E6 from puppet fileserver to user foo

```puppet
gnupg_key { 'jenkins_foo_key':
  ensure     => present,
  key_id     => 'D50582E6',
  user       => 'foo',
  key_source => 'puppet:///modules/gnupg/D50582E6.key',
  key_type   => public,
}
```

#### Add public key D50582E6 from puppet fileserver to user bar via a string value

```puppet
gnupg_key { 'jenkins_foo_key':
  ensure      => present,
  key_id      => 'D50582E6',
  user        => 'bar',
  key_content => '-----BEGIN BROKEN PUBLIC KEY BLOCK-----...',
  key_type    => public,
}
```
*Note*: You should use hiera lookup to get the key content

#### Remove public key 20BC0A86 from user root

```puppet
gnupg_key {'root_remove':
  ensure   => absent,
  key_id   => '20BC0A86',
  user     => 'root',
  key_type => public,
}
```

### Remove both private and public key 20BC0A66

```puppet
gnupg_key {'root_remove':
  ensure   => absent,
  key_id   => '20BC0A66',
  user     => 'root',
  key_type => both,
}
```

## Reference

See the [REFERENCE](REFERENCE.md)

### Tests

There are two types of tests distributed with the module. Unit tests with rspec-puppet and acceptance tests using puppet_litmus.

For unit testing, make sure you have:

* rake
* bundler

Install the necessary gems:

    bundle config set --local path 'vendor'
    bundle install

And then run the unit tests:

    bundle exec rake spec

If you want to run the system tests, make sure you also have:

* docker (default set)
* vagrant

Then run the tests using the default nodeset and puppet7:

    bundle exec rake 'litmus:provision_list[default]'
    bundle exec rake 'litmus:install_agent[puppet7]'
    bundle exec rake litmus:install_module

    bundle exec rake litmus:acceptance:parallel

Check the `provision.yaml` file for other sets you can use.

**Note**: Tests will start to fail after 2023-03-30 (jenkins key expires).
If this happens, please update the key used in the acceptance tests

## Limitations

This module has been tested on:

* Debian 8/9/10
* Ubuntu 14+
* RedHat 6/7/8
* CentOS 6/7/8

Note: For an up-to-date list, please check the acceptance tests.

## Development

Please see CONTRIBUTING.md

## License

See LICENSE file



CHANGELOG.md
2021-04-02 - Jan Vansteenkiste <jan@vstone.eu> - 2.0.1
* https and http support now both use the same method: fixes http->https redirects
* its difficult to find working upstream keyservers for the tests...

2021-04-01 - Jan Vansteenkiste <jan@vstone.eu> - 2.0.0
* Drop support for puppet < 6
* Merge multiple pull requests:
  * pr/35: Use ensure_packages to prevent conflicts
  * pr/33: Display the executed command on failure
  * pr/26: Allow full 40 char fingerprints  + Use batch import to prevent interactive prompts
* Conversion for acceptance tests to puppet litmus

2016-01-22 - Dejan Golja <dejan@golja.org> - 1.2.3
* Another retry to rebuild repack the module to fix the PaxHeaders bsd tar bug

2016-01-18 - Dejan Golja <dejan@golja.org> - 1.2.2
* module Remove PaxHeaders from distributed package, so it works with r10k

2015-07-04 - Dejan Golja <dejan@golja.org> - 1.2.1
* remove (present|absent) restriction on $package_ensure
* add support for SLES

2015-05-16 - Dejan Golja <dejan@golja.org> - 1.2.0
* autorequire user type when using gnupg_key (tnx to tcinel@atlassian.com)
* added two new custom facts:
  * gnupg_command - get full path to gpg exe
  * gnupg_installed - determine if gpg exe exists on system

2015-03-24 - Dejan Golja <dejan@golja.org> - 1.1.1
* supports username with -

2015-02-21 - Dejan Golja <dejan@golja.org> - 1.1.0
* properly supports private key (added new paramater key_type
  for gnupg_key type)
* added new parameter key_content for gnupg_key type
  to make it easier integrate hiera lookups. More info check README

Special thanks to Tobias Clemson (https://github.com/tobyclemson)

2014-10-20 - Dejan Golja <dejan@golja.org> - 1.0.0
* remove support for puppet 2.7.x due to the native type
* added beaker system spec tests
* Update type definition so it uses Puppet::Util::Execution.execute
  instead of Puppet::Util::SUIDManager.run_and_capture (deprecated)
* bump version to 1.0.0

2014-10-07 - Dejan Golja <dejan@golja.org> - 0.0.5
* bugfix for https keys

2014-08-17 - Dejan Golja <dejan@golja.org> - 0.0.4
* Update metadata.json and removed Modulefile due to 
  as of Puppet 3.6 the Modulefile has been deprecated

2014-03-16 - Dejan Golja <dejan@golja.org> - 0.0.3
* Added Puppet Forge Search by Module Compatibility

2013-10-22 - Dejan Golja <dejan@golja.org> - 0.0.2
* public release for Forge

2013-10-12 - Dejan Golja <dejan@golja.org> - 0.0.1
* initial beta version not released to the public


spec/acceptance/class_spec.rb
require 'spec_helper_acceptance'

describe 'gnupg class' do
  package_name = case os[:family].downcase
                 when 'redhat'
                   'gnupg2'
                 when 'debian'
                   'gnupg'
                 else
                   'gnupg'
                 end

  context 'default parameters' do
    it 'works with no errors' do
      pp = <<-EOS
      class { 'gnupg': }
      EOS

      # Run it twice and test for idempotency
      idempotent_apply(pp)
    end

    describe package(package_name) do
      it { is_expected.to be_installed }
    end
  end
end


spec/classes/gnupg_init_spec.rb
require 'spec_helper'

describe 'gnupg', type: :class do
  ['RedHat', 'Debian', 'Linux', 'Suse'].each do |system|
    if system == 'Linux'
      let(:facts) { { osfamily: 'Linux', operatingsystem: 'Amazon' } }
    else
      let(:facts) { { osfamily: system } }
    end

    it { is_expected.to contain_class('gnupg::install') }

    describe "gnupg on system #{system}" do
      context 'when enabled' do
        let(:params) do
          {
            package_ensure: 'present',
            package_name: 'gnupg',
          }
        end

        it {
          is_expected.to contain_package('gnupg').with('ensure' => 'present')
        }
      end

      context 'when disabled' do
        let(:params) do
          {
            package_ensure: 'absent',
            package_name: 'gnupg',
          }
        end

        it {
          is_expected.to contain_package('gnupg').with('ensure' => 'absent')
        }
      end
    end
  end
end


spec/spec_helper_acceptance_local.rb
# frozen_string_literal: true

require 'puppet_litmus'
require 'singleton'
require 'pathname'

class LitmusHelper
  include Singleton
  include PuppetLitmus
end

RSpec.configure do |c|
  c.formatter = :documentation
end

def gpg(gpg_cmd, options = { user: 'root' }, &block)
  user = options.delete(:user)
  gpg = "gpg #{gpg_cmd}"
  command("su #{user} -c \"#{gpg}\"", &block)
end

def transfer_file(file, destination)
  filepath = Pathname.new(file).absolute? ? file : File.expand_path(File.join(File.dirname(__FILE__), '..', file))
  content = File.read(filepath)
  write_file(content, destination)
end

RSpec::Matchers.define(:be_one_of) do |expected|
  match do |actual|
    expected.include?(actual)
  end

  failure_message do |actual|
    "expected one of #{expected}, got #{actual}"
  end
end


lib/puppet/provider/gnupg_key/gnupg.rb
require 'etc'
require 'tmpdir'
require 'puppet/file_serving/content'

Puppet::Type.type(:gnupg_key).provide(:gnupg) do
  @doc = 'Provider for gnupg_key type.'

  defaultfor kernel: 'Linux'
  confine kernel: 'Linux'

  def self.instances
    []
  end

  # although we do not use the commands class it's used to detect if the gpg and awk commands are installed on the system
  commands gpg: 'gpg'
  commands awk: 'awk'

  def gpgenv(resource)
    if resource[:gnupg_home]
      { 'GNUPGHOME' => resource[:gnupg_home] }
    else
      {}
    end
  end

  def remove_key
    begin
      fingerprint_command = "gpg --fingerprint --with-colons #{resource[:key_id]} | awk -F: '$1 == \"fpr\" {print $10;}'"
      fingerprint = Puppet::Util::Execution.execute(fingerprint_command, uid: user_id, custom_environment: gpgenv(resource))
    rescue Puppet::ExecutionFailure
      raise Puppet::Error, "Could not determine fingerprint for  #{resource[:key_id]} for user #{resource[:user]}: #{fingerprint}"
    end

    if resource[:key_type] == :public
      command = "gpg --batch --yes --delete-key #{fingerprint}"
    elsif resource[:key_type] == :private
      command = "gpg --batch --yes --delete-secret-key #{fingerprint}"
    elsif resource[:key_type] == :both
      command = "gpg --batch --yes --delete-secret-and-public-key #{fingerprint}"
    end

    begin
      output = Puppet::Util::Execution.execute(command, uid: user_id, custom_environment: gpgenv(resource))
    rescue Puppet::ExecutionFailure
      raise Puppet::Error, "Could not remove #{resource[:key_id]} for user #{resource[:user]}: #{output}"
    end
  end

  # where most of the magic happens
  # TODO implement dry-run to check if the key_id match the content of the file
  def add_key
    if resource[:key_server]
      add_key_from_key_server
    elsif resource[:key_source]
      add_key_from_key_source
    elsif resource[:key_content]
      add_key_from_key_content
    end
  end

  def add_key_from_key_server
    command = if resource[:proxy].nil? || resource[:proxy].empty?
                "gpg --keyserver #{resource[:key_server]} --recv-keys #{resource[:key_id]}"
              else
                "gpg --keyserver #{resource[:key_server]} --keyserver-options http-proxy=#{resource[:proxy]} --recv-keys #{resource[:key_id]}"
              end
    begin
      Puppet::Util::Execution.execute(command, uid: user_id, failonfail: true, custom_environment: gpgenv(resource))
    rescue Puppet::ExecutionFailure => e
      raise Puppet::Error, "Key #{resource[:key_id]} does not exist on #{resource[:key_server]}\n#{e}"
    end
  end

  def add_key_from_key_source
    if Puppet::Util.absolute_path?(resource[:key_source])
      add_key_at_path
    else
      add_key_at_url
    end
  end

  def add_key_from_key_content
    path = create_temporary_file(user_id, resource[:key_content])
    command = "gpg --batch --import #{path}"
    begin
      output = Puppet::Util::Execution.execute(command, uid: user_id, failonfail: true, custom_environment: gpgenv(resource))
    rescue Puppet::ExecutionFailure
      raise Puppet::Error, "Error while importing key #{resource[:key_id]} using key content:\n#{output}}"
    end
  end

  def add_key_at_path
    unless File.file?(resource[:key_source])
      raise Puppet::Error, "Local file #{resource[:key_source]} for #{resource[:key_id]} does not exists"
    end
    command = "gpg --batch --import #{resource[:key_source]}"
    begin
      Puppet::Util::Execution.execute(command, uid: user_id, failonfail: true, custom_environment: gpgenv(resource))
    rescue Puppet::ExecutionFailure
      raise Puppet::Error, "Error while importing key #{resource[:key_id]} from #{resource[:key_source]}"
    end
  end

  def add_key_at_url
    uri = URI.parse(URI.escape(resource[:key_source]))
    case uri.scheme
    when %r{http}
      command = "wget -O- #{resource[:key_source]} | gpg --batch --import"
    when 'puppet'
      path = create_temporary_file user_id, puppet_content
      command = "gpg --batch --import #{path}"
    end
    begin
      output = Puppet::Util::Execution.execute(command, uid: user_id, failonfail: true, custom_environment: gpgenv(resource))
      if output =~ %r{unable to fetch}
        raise Puppet::ExecutionFailure, output
      end
    rescue Puppet::ExecutionFailure => e
      raise Puppet::Error, "Error while importing key #{resource[:key_id]} from #{resource[:key_source]}:\n#{e}"
    end
  end

  def user_id
    Etc.getpwnam(resource[:user]).uid
  rescue
    raise Puppet::Error, "User #{resource[:user]} does not exists"
  end

  def create_temporary_file(user_id, content)
    Puppet::Util::SUIDManager.asuser(user_id) do
      tmpfile = Tempfile.open(['golja-gnupg', 'key'])
      tmpfile.write(content)
      tmpfile.flush
      break tmpfile.path.to_s
    end
  end

  def puppet_content
    # Look up (if necessary) and return remote content.
    return @content if @content
    file_resource = Puppet::FileServing::Content.indirection.find(resource[:key_source], environment: resource.catalog.environment, links: :follow)
    unless file_resource
      raise 'Could not find any content at %s' % resource[:key_source]
    end
    @content = file_resource.content
  end

  def exists?
    # public and both can be grouped since private can't be present without public,
    # both only applies to delete and delete still has something to do if only
    # one of the keys is present
    if resource[:key_type] == :public || resource[:key_type] == :both
      command = "gpg --list-keys --with-colons #{resource[:key_id]}"
    elsif resource[:key_type] == :private
      command = "gpg --list-secret-keys --with-colons #{resource[:key_id]}"
    end

    output = Puppet::Util::Execution.execute(command, uid: user_id, custom_environment: gpgenv(resource))
    return true if output.exitstatus.zero?
    return false if output.exitstatus == 2
    raise Puppet::Error, "Non recognized exit status from GnuPG #{output.exitstatus} #{output}"
  end

  def create
    add_key
  end

  def destroy
    remove_key
  end
end


lib/puppet/type/gnupg_key.rb
require 'uri'
Puppet::Type.newtype(:gnupg_key) do
  @doc = 'Manage PGP public keys with GnuPG'

  ensurable

  autorequire(:package) do
    ['gnupg', 'gnupg2']
  end

  autorequire(:user) do
    self[:user]
  end

  KEY_SOURCES = [:key_source, :key_server, :key_content].freeze

  KEY_CONTENT_REGEXES = {
    public: ['-----BEGIN PGP PUBLIC KEY BLOCK-----', '-----END PGP PUBLIC KEY BLOCK-----'],
    private: ['-----BEGIN PGP PRIVATE KEY BLOCK-----', '-----END PGP PRIVATE KEY BLOCK-----'],
  }.freeze

  validate do
    creator_count = 0
    KEY_SOURCES.each do |param|
      creator_count += 1 unless self[param].nil?
    end

    if creator_count > 1
      raise ArgumentError, "You cannot specify more than one of #{KEY_SOURCES.map { |p| p.to_s }.join(', ')}, " \
                           'much to learn, you still have.'
    end

    if creator_count.zero? && self[:ensure] == :present
      raise ArgumentError, "You need to specify at least one of #{KEY_SOURCES.map { |p| p.to_s }.join(', ')}, " \
                           'much to learn, you still have.'
    end

    if self[:ensure] == :present && self[:key_type] == :both
      raise ArgumentError, "A key type of 'both' is invalid when ensure is 'present'."
    end

    [:public, :private].each do |type|
      next unless self[:key_content] && self[:key_type] == type
      key_lines = self[:key_content].strip.lines.to_a

      first_line = key_lines.first.strip
      last_line = key_lines.last.strip

      unless first_line == KEY_CONTENT_REGEXES[type][0] && last_line == KEY_CONTENT_REGEXES[type][1]
        raise ArgumentError, "Provided key content does not look like a #{type} key."
      end
    end
  end

  newparam(:name, namevar: true) do
    desc "This attribute is currently used as a
      system-wide primary key - namevar and therefore has to be unique."
  end

  newparam(:user) do
    desc "The user account in which the PGP public key should be installed.
    Usually it's stored in HOME/.gnupg/ dir"

    validate do |value|
      # freebsd/linux username limitation
      unless value =~ %r{^[a-z_][a-z0-9_-]*[$]?}
        raise ArgumentError, "Invalid username format for #{value}"
      end
    end
  end

  newparam(:key_source) do
    desc <<-'EOT'
      A source file containing PGP key. Values can be URIs pointing to remote files,
      or fully qualified paths to files available on the local system.

      The available URI schemes are *puppet*, *https*, *http* and *file*. *Puppet*
      URIs will retrieve files from Puppet's built-in file server, and are
      usually formatted as:

      `puppet:///modules/name_of_module/filename`
    EOT

    validate do |source|
      raise ArgumentError, 'Arrays not accepted as an source parameter' if source.is_a?(Array)
      break if Puppet::Util.absolute_path?(source)

      begin
        uri = URI.parse(source)
      rescue => detail
        raise ArgumentError, "Could not understand source #{source}: #{detail}"
      end

      raise ArgumentError, "Cannot use relative URLs '#{source}'" unless uri.absolute?
      raise ArgumentError, "Cannot use opaque URLs '#{source}'" unless uri.hierarchical?
      raise ArgumentError, "Cannot use URLs of type '#{uri.scheme}' as source for fileserving" unless ['file', 'puppet', 'https', 'http'].include?(uri.scheme)
    end

    munge do |source|
      if ['file'].include?(URI.parse(source).scheme)
        URI.parse(source).path
      else
        source
      end
    end
  end

  newparam(:key_server) do
    desc 'PGP key server from where to retrieve the public key'

    validate do |server|
      if server
        uri = URI.parse(server)
        unless uri.is_a?(URI::HTTP) || uri.is_a?(URI::HTTPS) ||
               uri.is_a?(URI::LDAP) || ['hkp'].include?(uri.scheme)
          raise ArgumentError, "Invalid keyserver value #{server}"
        end
      end
    end
  end

  newparam(:key_content) do
    desc "Key content. The result of exporting the key using ASCII armor.
      Can be either a public or private key."
  end

  newparam(:key_id) do
    desc "Key ID. Usually the traditional 8-character key ID. Also accepted the
      long more accurate (but  less  convenient) 16-character key ID."

    validate do |value|
      unless ([8, 16, 40].include? value.length) && value =~ %r{^[0-9A-Fa-f]+$}
        raise ArgumentError, "Invalid key id #{value}"
      end
    end

    munge do |value|
      value.upcase.to_sym
    end
  end

  newparam(:key_type) do
    desc 'The type of the key(s) being managed.'

    newvalues(:public, :private, :both)

    defaultto :public
  end

  newparam(:proxy) do
    desc 'Set the proxy to use for HTTP and HKP keyservers.'

    validate do |value|
      if value
        uri = URI.parse(value)
        unless uri.is_a?(URI::HTTP) || uri.is_a?(URI::HTTPS)
          raise ArgumentError, "Invalid proxy value #{value}"
        end
      end
    end
  end

  newparam(:gnupg_home) do
    desc "The gnupg home directory. Overrides the default user's homedir."
    defaultto false

    validate do |value|
      unless Puppet::Util.absolute_path?(value)
        raise Puppet::Error, _("File paths must be fully qualified, not '%{path}'") % { path: value }
      end
    end

    munge do |value|
      ::File.join(::File.split(::File.expand_path(value)))
    end
  end
end
